model/riscv_insts_zfa.sail

```diff
@@ -0,0 +1,770 @@
+               ^ spc() ^ freg_name(rs2)
+
+function clause execute(RISCV_FLEQ_S(rs2, rs1, rd)) = {
+  let rs1_val_S = F_S(rs1);
+  let rs2_val_S = F_S(rs2);
+
+  let (fflags, rd_val) : (bits_fflags, bool) =
+      riscv_f32Le_quiet   (rs1_val_S, rs2_val_S);
+
+  write_fflags(fflags);
+  X(rd) = EXTZ(bool_to_bits(rd_val));
+  RETIRE_SUCCESS
+}
+
+/* FLTQ.S */
+
+union clause ast = RISCV_FLTQ_S : (regidx, regidx, regidx)
+
+mapping clause encdec = RISCV_FLTQ_S(rs2, rs1, rd)               if haveZfa()
+  <-> 0b101_0000 @ rs2 @ rs1 @ 0b101 @ rd @ 0b101_0011           if haveZfa()
+
+mapping clause assembly = RISCV_FLTQ_S(rs2, rs1, rd)
+  <-> "fltq.s" ^ spc() ^ freg_name(rd)
+               ^ spc() ^ freg_name(rs1)
+               ^ spc() ^ freg_name(rs2)
+
+function clause execute(RISCV_FLTQ_S(rs2, rs1, rd)) = {
+  let rs1_val_S = F_S(rs1);
+  let rs2_val_S = F_S(rs2);
+
+  let (fflags, rd_val) : (bits_fflags, bool) =
+      riscv_f32Lt_quiet   (rs1_val_S, rs2_val_S);
+
+  write_fflags(fflags);
+  X(rd) = EXTZ(bool_to_bits(rd_val));
+  RETIRE_SUCCESS
+}
+
+
+/* FLEQ.D */
+
+union clause ast = RISCV_FLEQ_D : (regidx, regidx, regidx)
+
+mapping clause encdec = RISCV_FLEQ_D(rs2, rs1, rd)               if (haveDExt() & haveZfa())
+  <-> 0b101_0001 @ rs2 @ rs1 @ 0b100 @ rd @ 0b101_0011           if (haveDExt() & haveZfa())
+
+mapping clause assembly = RISCV_FLEQ_D(rs2, rs1, rd)
+  <-> "fleq.d" ^ spc() ^ freg_name(rd)
+               ^ spc() ^ freg_name(rs1)
+               ^ spc() ^ freg_name(rs2)
+
+function clause execute(RISCV_FLEQ_D(rs2, rs1, rd)) = {
+  let rs1_val_D = F_D(rs1);
+  let rs2_val_D = F_D(rs2);
+
+  let (fflags, rd_val) : (bits_fflags, bool) =
+      riscv_f64Le_quiet   (rs1_val_D, rs2_val_D);
+
+  write_fflags(fflags);
+  X(rd) = EXTZ(bool_to_bits(rd_val));
+  RETIRE_SUCCESS
+}
+
+/* FLTQ.D */
+
+union clause ast = RISCV_FLTQ_D : (regidx, regidx, regidx)
+
+mapping clause encdec = RISCV_FLTQ_D(rs2, rs1, rd)               if (haveDExt() & haveZfa())
+  <-> 0b101_0001 @ rs2 @ rs1 @ 0b101 @ rd @ 0b101_0011           if (haveDExt() & haveZfa())
+
+mapping clause assembly = RISCV_FLTQ_D(rs2, rs1, rd)
+  <-> "fltq.d" ^ spc() ^ freg_name(rd)
+               ^ spc() ^ freg_name(rs1)
+               ^ spc() ^ freg_name(rs2)
+
+function clause execute(RISCV_FLTQ_D(rs2, rs1, rd)) = {
+  let rs1_val_D = F_D(rs1);
+  let rs2_val_D = F_D(rs2);
+
+  let (fflags, rd_val) : (bits_fflags, bool) =
+      riscv_f64Lt_quiet   (rs1_val_D, rs2_val_D);
+
+  write_fflags(fflags);
+  X(rd) = EXTZ(bool_to_bits(rd_val));
+  RETIRE_SUCCESS
+}
+
+/* FCVTMOD.W.D */
+
+/*
+ * Implement float64 to int32 conversion without saturation, the
+ * result is supplied modulo 2^32.
+ * Rounding mode for FCVTMOD.W.D is always RTZ(Round toward zero).
+ */
+val      fcvtmod_helper : forall 'm, 'm == 64. (bits('m)) -> (bits(5), xlenbits)
+function fcvtmod_helper(x64) = {
+  let (sign, exp, mant) = fsplit_D(x64);
+  fflags		= fcsr.FFLAGS();
+  result : bits(64)	= undefined;
```

Comment:
Hm. Changing this is not as straightforward as it looked. I'll take another look in the coming days to see if this can be expressed more cleanly with a major rewrite of the function.

Reply:
Mind if we kick this can down the road (and deal with it when we reimplement softfloat in pure-SAIL)?

I looked at it and it will be a significant rewrite of the function as we depend on the checks at the end of the if block as a common path...

Reply:
We plan to have a rewritten version available by end of this week.

Reply:
@jrtc27 Could you take a peek at the reeimplementatio and let us know if this is good SAIL style (or rather: what should be changed)?

Reply:
@jrtc27 Any thoughts on this?

Reply:
Managed to miss your previous comment; I'll take a look later tonight

Reply:
Thank you!